### ISSUE TITLE
Analyzer that forbids the use of struct default c'tors

### ISSUE THREAD

**Issue by lachbaer** [dotnet/roslyn-analyzers#1185]
I'm referring to https://github.com/dotnet/csharplang/issues/99#issuecomment-298840503. @CyrusNajmabadi sent me here.

To be (mostly) sure that a `struct` is initialized correctly, independently from https://github.com/dotnet/csharplang/issues/99, it would be good to ensure that `default(S)` or `new S()` is not used in (i.a. ones own) user code. Instead a custom parametered c'tor of the `struct` should be used everytime.

I think that it would be good for that case to have an attribute, e.g. "[ConstructorMandatory]", that lets the analyzer check for the (public) existence of a parametered c'tor in the `struct` and ensure that `default(S)` and `new S()` are never used for this struct, except in the corresponding struct itself.

Of course that doesn't prevent from getting uninitialized instances, but at least it can prevent from accidentially creating one directly.

The analyzer should be enabled by default, because it only inspects tagged structs and has no impact on existing code.

The docs for the analyzer shall state that the parametered c'tor can be a "dummy":
```cs
public Fraction(object dummy) { Denominator = 1d; }
public Fraction(Fraction? other) { 
    this = other.HasValue ? other.Value : new Fraction() { Denominator = 1d }; 
}

var frac = new Fraction(null);    // instead of `new Fraction();`
```

**Comment by sharwell** [comment]
What about the following cases?

1. The value returned from `IDictionary.TryGetValue` when a key is not found
2. Using the struct as the argument for a generic parameter with the `new()` constraint
3. Using the struct as the argument for a generic parameter, where the implementation uses `default(T)` to obtain a default instance of the type argument
4. Creating arrays

I agree that `new S()` should not be used when `S` is a value type. If in the future parameterless user-defined constructors for value types are allowed, then this rule would be restricted to cases where the type does not have such a constructor. I do not agree with the same applying to `default(S)`.

This is related to DotNetAnalyzers/StyleCopAnalyzers#1482.

**Comment by sharwell** [comment]
💭 It would be possible to create an analyzer for this provided it had the following additional characteristics:

1. The analyzer _only_ looks at types with a particular attribute, e.g. `[ConstructorMandatory]`.
2. The analyzer reports a warning for instances of `default(S)` and `new S()`.
3. The analyzer reports a warning for any attempted use of `S` as a generic type parameter (users must instead use `S?`).
4. The analyzer reports a warning for any attempted use of `S` to create an array, e.g. `new S[expression]` (users must instead use `new S?[expression]`).

**Comment by lachbaer** [comment]
>I do not agree with the same applying to default(S)

Of course an analyzer should be configurable. The intend to forbid non-generic `default(S)` also is that the user is forced to use a parametered constructor and thus nearly never can produce an invalid, uninitialized value instance.

**Comment by Evangelink** [comment]
> I would recommend this analyzer be implemented independently, perhaps as part of a special-purpose NuGet package that also provides the definition for the `[ConstructorMandatory]` attribute.

@sharwell Why do you want to prevent the analyzer to be part of the regular package? Regarding the attribute there seems to have some discussion going-on.

**Comment by sharwell** [comment]
My focus for this feature right now is on developing the behavior described in https://github.com/dotnet/csharplang/issues/99#issuecomment-601792573
